hosted/aks/helper/helper_cluster.go

```diff
@@ -28,46 +32,140 @@ var (
-func UpgradeClusterKubernetesVersion(cluster *management.Cluster, upgradeToVersion *string, client *rancher.Client) (*management.Cluster, error) {
+func UpgradeClusterKubernetesVersion(cluster *management.Cluster, upgradeToVersion string, client *rancher.Client, check bool) (*management.Cluster, error) {
+	currentVersion := *cluster.AKSConfig.KubernetesVersion
 	upgradedCluster := new(management.Cluster)
 	upgradedCluster.Name = cluster.Name
 	upgradedCluster.AKSConfig = cluster.AKSConfig
-	upgradedCluster.AKSConfig.KubernetesVersion = upgradeToVersion
+	upgradedCluster.AKSConfig.KubernetesVersion = &upgradeToVersion
 
-	cluster, err := client.Management.Cluster.Update(cluster, &upgradedCluster)
+	var err error
+	cluster, err = client.Management.Cluster.Update(cluster, &upgradedCluster)
 	if err != nil {
 		return nil, err
 	}
+
+	if check {
+		// Check if the desired config is set correctly
+		Expect(*cluster.AKSConfig.KubernetesVersion).To(Equal(upgradeToVersion))
+		// ensure nodepool version is still the same when config is applied
+		for _, np := range cluster.AKSConfig.NodePools {
+			Expect(*np.OrchestratorVersion).To(Equal(currentVersion))
+		}
+
+		// Check if the desired config has been applied in Rancher
+		Eventually(func() string {
+			ginkgo.GinkgoLogr.Info("Waiting for k8s upgrade to appear in AKSStatus.UpstreamSpec ...")
+			cluster, err = client.Management.Cluster.ByID(cluster.ID)
+			Expect(err).NotTo(HaveOccurred())
+			return *cluster.AKSStatus.UpstreamSpec.KubernetesVersion
+		}, tools.SetTimeout(10*time.Minute), 5*time.Second).Should(Equal(upgradeToVersion))
+		// ensure nodepool version is same in Rancher
+		for _, np := range cluster.AKSStatus.UpstreamSpec.NodePools {
+			Expect(*np.OrchestratorVersion).To(Equal(currentVersion))
+		}
+
+	}
 	return cluster, nil
 }
 
 // UpgradeNodeKubernetesVersion upgrades the k8s version of nodepool to the value defined by upgradeToVersion.
-func UpgradeNodeKubernetesVersion(cluster *management.Cluster, upgradeToVersion *string, client *rancher.Client) (*management.Cluster, error) {
+func UpgradeNodeKubernetesVersion(cluster *management.Cluster, upgradeToVersion string, client *rancher.Client, wait, check bool) (*management.Cluster, error) {
 	upgradedCluster := new(management.Cluster)
 	upgradedCluster.Name = cluster.Name
 	upgradedCluster.AKSConfig = cluster.AKSConfig
 	for i := range upgradedCluster.AKSConfig.NodePools {
-		upgradedCluster.AKSConfig.NodePools[i].OrchestratorVersion = upgradeToVersion
+		upgradedCluster.AKSConfig.NodePools[i].OrchestratorVersion = &upgradeToVersion
 	}
 
 	cluster, err := client.Management.Cluster.Update(cluster, &upgradedCluster)
 	if err != nil {
 		return nil, err
 	}
+
+	if check {
+		// Check if the desired config is set correctly
+		for _, np := range cluster.AKSConfig.NodePools {
+			Expect(*np.OrchestratorVersion).To(Equal(upgradeToVersion))
+		}
+	}
+
+	if wait {
```

Comment:
if wait, check = false, how to assert test result ?
Since we have added one more level of validation, check can be true by default IMO

Reply:
If they are false, it is assumed that the assertion and wait will be done in the test, check k8s_chart_support_upgrade for eg.